Submission implementation

atarix83 opened the pull request:
This PR implements Submission functionalities and contains fixes included in [PR 219](https://github.com/DSpace/dspace-angular/pull/219)

These REST PRs need to be merged before we can merge this one:
- https://github.com/DSpace/DSpace/pull/2102
- https://github.com/DSpace/DSpace/pull/2103

This implementation allow : 

- to create a new submission by the route `/submit`
- to edit an existing submission by the route `workspaceitems/<id>/edit`
- to edit a submission already deposited by the route `workflowitems/<id>/edit`

For testing new functionalities please use the OR2018 REST server by changing the default rest namespace configuration to `/or2018/api`.

Unit tests are still missing.

AlexanderS commented:
Wow, that's a lot of code in only one commit.

tdonohue commented:
I gave this a test today by pulling down the PR and setting the `rest.nameSpace = '/or2018/api'` in my `environment.default.js`.  I then logged in as "dspacedemo+submit@gmail.com" using the same password as on demo.dspace.org

👍 The new `/submit` path seems to work well.  I tried saving an in-progress submission for later as well as submitting a new item.  Both options seem to work.  I also tried leaving out or incorrectly entering various submission fields to ensure validation was working -- and it seems to be.

However, without access to the backend server (or more feedback from the UI), I cannot easily test the `workspaceitems/<id>/edit` or ` workflowitems/<id>/edit` paths. I don't seem to be given any `<id>` for submissions that I save or submit to a workflow-enabled collection (since the "MyDSpace" UI hasn't been completed yet).   Could we perhaps generate a few test workspace/workflow items (on the backend) that can be used for testing these new interfaces?

(As a sidenote, I have not yet done a code review of any of this work. Just wanted to try a simple test)

atarix83 commented:
> However, without access to the backend server (or more feedback from the UI), I cannot easily test the workspaceitems/<id>/edit or workflowitems/<id>/edit paths. I don't seem to be given any <id> for submissions that I save or submit to a workflow-enabled collection (since the "MyDSpace" UI hasn't been completed yet). Could we perhaps generate a few test workspace/workflow items (on the backend) that can be used for testing these new interfaces?

With HAL browser it's possible to list workspaceitems and workflowitems and use one of those : 
https://dspace7.4science.it/or2018/#https://dspace7.4science.it/or2018/api/submission/workspaceitems
https://dspace7.4science.it/or2018/#https://dspace7.4science.it/or2018/api/workflow/workflowitems

While if you'd be sure to edit a new workspaceitems created, during submission you can open your browser's console and in the network section watching the REST call that contains the ID of the workspaceitems 
![schermata da 2018-08-31 10-16-13](https://user-images.githubusercontent.com/2486489/44902073-4f353c80-ad09-11e8-862a-9332f7a13167.jpg)

atarix83 commented:
@artlowel 
>I submitted this item (source), and I noticed that when I add the third identifier, the automatic metadata extraction is called for the second time and as a result all automatically extracted fields are added twice. I was able to reproduce the issue reliably for this item, and also for this pdf (didn't actually submit that one as to not pollute your demo api too much)

Automatic metadata extraction is related to the REST server configuration. 
In case of OR2018 server configuration, metadata extraction should be called when uploading a PDF file, a BibTeX file, or adding a DOI identifier. If you add a DOI identifier related a pdf previously added to the submission, metadata extraction is called again, but only fields for which extraction provide different values should be added many times to the input form. Then the interface allows the user to delete the wrong value.

However there was a bug with identifiers field fixed with last commit.
Could you check if the you can reproduce the issue again, If there is still, could you tell me which values of identifiers you use? thanks

atarix83 commented on src/app/shared/services/route.service.ts:
```diff
@@ -40,4 +39,21 @@ export class RouteService {
+    return this.history;
+  }
+
+  public getPreviousUrl(): string {
```
I can't find navigations actions in the master version of documentation, but I found them for v5 here: 
https://github.com/ngrx/platform/blob/v5.2.0/docs/router-store/api.md#navigation-actions

However it seems that navigation actions are not provided as part of the router package, but they can be added as described in the documentation.
This implementation use `Location` object to navigate back or forward.

Do you think that is the best solution?

artlowel replied:
Hmm, they removed those docs about navigation actions in [this pr](https://github.com/ngrx/platform/pull/1283) a few days after I made that comment.

You're right, it seems like [Location.back()](https://angular.io/api/common/Location#back) is now the way to go, that doesn't require tracking the history.

atarix83 replied:
ok Location.back() works, but I realized that I need the previous url to make a check and then navigate to it or not, so I'd require tracking the history

artlowel commented:
>Could you check if the you can reproduce the issue again, If there is still, could you tell me which values of identifiers you use? thanks

I'm still testing with [this paper](https://www.frontiersin.org/articles/10.3389/fpsyg.2013.00479/full), 
and it happens when I add the doi (10.3389/fpsyg.2013.00479) and then wait until it's done saving. 

Now only the title and abstract are added twice, and indeed, there are very slight differences between the fields. One title has a full stop behind it, the other doesn't, one abstract uses an em-dash somewhere in the middle, the other uses a regular dash.

artlowel commented on src/app/core/eperson/eperson.service.ts:
```diff
@@ -0,0 +1,53 @@
+import { HALEndpointService } from '../shared/hal-endpoint.service';
+import { EpersonData } from './eperson-data';
+
+export abstract class EpersonService {
```
>There is a specific reason for the presence of this filter?

Yes, if the response isn't successful there will be no pageInfo object, and the code after it won't work

I think the cause of this issue is that the rest api returns a 404. If there are no members of a group you'd expect an empty page, like you see here for example: https://dspace7.4science.it/or2018/#https://dspace7.4science.it/or2018/api/core/communities/search/subCommunities?parent=7669c72a-3f2a-451f-a3b9-9210e7a4c02f 

404 should imo only be thrown if the endpoint doesn't exist, or the user isn't authorized to use it (and simply knowing of the existence of the endpoint would give away sensitive information). 

What do you think, @abollini ?

abollini replied:
I agree, it is better to fix the REST API implementation returning an "empty page" (200) or eventually a nocontent.
Conceptually I tend to prefer 200 with an empty list or result and proper pagination info over the nocontent (204) solution.
On the other hand the nocontent solution is probably much easier to enforce centrally in the RESTResourceController, see https://github.com/DSpace/DSpace/pull/2189

atarix83 commented:
@tdonohue @artlowel 

PR is ready to new reviews

tdonohue commented:
@atarix83 : Glad to see you got this updated!  Just a quick question, are unit tests / specs still being worked on?  I notice that coverage decreases by 11% with this PR (goes from 76% coverage down to 65%).  Obviously, that's a massive drop.

In any case, I'll give this an updated test and report back on results. Thanks!

atarix83 commented:
@tdonohue unit tests are work in progress, but I don't know how long they take

However when the following REST side PR will have been merged and installed on the demo, it'll be possible to test this PR with REST demo server ([https://dspace7.4science.it](https://dspace7.4science.it)) : 
[DSpace/DSpace#2239](https://github.com/DSpace/DSpace/pull/2239)

artlowel commented:
I didn't have the time yet to go over the code in detail again. But I did test the UI and came across a few bugs. I'm running `yarn run start`, against `https://dspace7.4science.it/or2018/api`. I was logged in for all tests:
- In general
	- Collection pages are broken: 
		- The license is rendered as `[object Object]`
		- The recent submissions don't load: Error message: `Cannot read property 'objects' of undefined`
	- Search page is broken as well. Errors:
		- `Cannot read property '_embedded' of undefined`
		- `Cannot read property 'objects' of undefined`
- On the submission page
	- In the Collection dropdown:
		- If you type a query, and remove it again, the remove doesn't trigger an update of the results until you click outside of the dropdown (at which point it disappears)
	- In the Authors field:
		- if you enter an author name, click the undo button, and paste it in again, the field turns red and says it is required, even though it's filled in.




I also came across these usability issues:
- In the Authors field, it took me some trial and error to figure out:
	- that you're supposed to enter a single author, and then click the save icon. I think having the return key (form submit) trigger a save would help in that regard.
	- The way the trashcan works: it doesn't do anything until you've saved. I expected it to clear what I've typed
	- The way the undo button works: It _does_ clear what you've typed before you've saved. It only becomes clear why it's an _undo_ button after you've saved, and started editing again. Perhaps hide it until after you've saved?
	- How to add additional authors. I'd expect a plus sign or the word "add" rather than a downward chevron.
- There are 2 other ways of adding multivalued fields: the keywords way, where they turn into labels as soon as the input loses focus and sponsors, with add and remove buttons. Both seem more intuitive than the way it works for authors. I think we should choose one way of doing this and stick to it.
- While entering metadata, you often get a warning saying there are incomplete fields. It would be better if that warning only came when you finish (try to deposit)

artlowel commented on src/app/submission/submission.service.spec.ts:
```diff
@@ -355,15 +359,24 @@ describe('SubmissionService test suite', () => {
         { provide: Router, useValue: router },
         { provide: SubmissionRestService, useValue: restService },
         { provide: ActivatedRoute, useValue: new MockActivatedRoute() },
+        NotificationsService,
```
Please try to mock any external services you use in tests. See https://github.com/DSpace/dspace-angular/issues/281
If we use actual services, the tests not only become harder to write, but they can break when the external service changes. They can also influence and break other, seemingly unrelated tests, causing issues that are very hard to debug. (e.g. tests that fail only when a specific other, unrelated test has run first)

tdonohue commented:
@atarix83 : I gave this a test today, and I'm running into problems similar to what @artlowel reported above. Additionally, I've noticed:

Using the default REST API (https://dspace7.4science.it/dspace-spring-rest/api/)  unfortunately does **not** entirely work (even though that REST API has been updated to include all of @abollini's recent PRs related to submission).  
* Search pages work properly.  Browsing also works properly.
* Collection pages "mostly" work, but I do see an odd `[object Object]` message in the License section.
* When visiting `/submit`, even if I'm already logged in, I'm prompted to login a second time.
* On the http://localhost:3000/submit page, The list of collections never successfully loads 
    * The first "Describe" box loads properly, and I'm able to enter values in the metadata fields. But, I can never get a green check to appear, even after entering all required fields.
    * In the second "Describe" box, I see this error: `An error occurred during section initialize, please check your input-form configuration. Details are below : Error: unknown form control model type "twobox" defined for Input field with label "Subject Keywords".`
    * Uploading files does not work. (The file appears to upload but never appear in the "Upload files" section of the form)
    * None of the buttons at the bottom work (Save, Save for Later or Deposit).

Using the OR2018 branch REST API (https://dspace7.4science.it/or2018/api/) allows the http://localhost:3000/submit page to load all its data properly *except for the Collection listing* (which is still stuck in a loading phase).  I'm able to fill out metadata in the submission form, but I'm still unable to upload any files (they appear to upload, but never appear in the "Upload files" section of the form).
* As @artlowel noted, in this scenario, the Collection and Search pages do not appear to work anymore.  My best guess is that this older branch doesn't include recent REST API updates which causes these failures

tdonohue commented on src/app/core/data/remote-data-error.ts:
```diff
@@ -1,6 +1,6 @@
 export class RemoteDataError {
   constructor(
-    public statusCode: string,
+    public statusText: string,
```
I realize I forgot to mention... in testing this PR yesterday, I discovered that this `RemoteDataError` change breaks the `search-results` component on `master`. Unfortunately, it currently depends on / uses the old `statusCode` property here: https://github.com/DSpace/dspace-angular/blob/master/src/app/+search-page/search-results/search-results.component.html#L10-L11

Hopefully we can find a way to resolve this, but this is the cause of (at least some) Search Results problems

atarix83 commented on src/app/submission/form/collection/submission-form-collection.component.html:
```diff
@@ -0,0 +1,38 @@
+<div>
+  <div ngbDropdown #collectionControls="ngbDropdown" class="btn-group input-group" (openChange)="toggled($event)">
+    <div class="input-group-prepend">
+      <span id="collectionControlsMenuLabel" class="input-group-text">Collection</span>
```
Fixed

atarix83 commented on src/app/submission/form/collection/submission-form-collection.component.html:
```diff
@@ -0,0 +1,38 @@
+      </div>
+      <div class="dropdown-divider"></div>
+      <div class="scrollable-menu" aria-labelledby="dropdownMenuButton" (scroll)="onScroll($event)">
+        <button class="dropdown-item disabled" *ngIf="(searchListCollection$ | async).length == 0">No results found</button>
```
Fixed

atarix83 commented on src/app/submission/form/section-add/submission-form-section-add.component.html:
```diff
@@ -0,0 +1,14 @@
+          {{'submission.sections.general.add-more' | translate}} <i class="fa fa-plus" aria-hidden="true"></i>
+  </button>
+  <div ngbDropdownMenu class="sections-dropdown-menu" aria-labelledby="sectionControls" [ngClass]="{'w-100': (windowService.isXs() | async)}">
+    <button class="dropdown-item disabled" *ngIf="(sectionList | async)?.length == 0">No options available</button>
```
Fixed

atarix83 commented on src/app/submission/sections/upload/section-upload.component.html:
```diff
@@ -0,0 +1,50 @@
+<ng-container *ngIf="fileList.length == 0">
+  <div class="row">
+    <div class="col-md-12">
+      <h3 class="text-center"><span class="text-muted">No file uploaded yet.</span></h3>
```
Fixed

atarix83 commented on src/app/submission/sections/upload/file/view/file-view.component.html:
```diff
@@ -0,0 +1,25 @@
+          <span class="text-muted">No {{entry.key}}</span>
+        </h5>
+        <ng-container *ngIf="metadata.indexOf(entry) !== 0">
+          <span class="text-muted">No {{entry.key}}</span>
```
Fixed

atarix83 commented on src/app/core/json-patch/json-patch.model.ts:
```diff
@@ -0,0 +1,14 @@
+  op: JsonPatchOperationType;
+  path: string;
+  value: any;
+}
```
Fixed

atarix83 commented on src/app/submission/edit/submission-edit.component.ts:
```diff
@@ -0,0 +1,75 @@
+        this.submissionId = params.get('id');
+        this.subs.push(
+          this.submissionService.retrieveSubmission(this.submissionId)
+            .subscribe((submissionObject: SubmissionObject) => {
```
Fixed

atarix83 commented on src/app/shared/form/builder/models/form-field-metadata-value.model.ts:
```diff
@@ -1,39 +1,47 @@
   place: number;
   closed: boolean;
   label: string;
+  otherInformation: any;
```
I used same nomenclature used on REST side, I've added a type for this property

atarix83 commented on src/app/core/integration/models/authority.value.ts:
```diff
@@ -0,0 +1,42 @@
+  value: string;
+
+  @autoserialize
+  otherInformation: any;
```
same as above

tdonohue commented on src/app/core/cache/models/search-param.model.ts:
```diff
@@ -0,0 +1,6 @@
+
+export class SearchParam {
```
Please add TypeDoc/comments to describe this new class.

Is this class purposefully empty?  It looks a little odd. But, maybe it'll be easier to understand if you add a comment to describe it.

tdonohue commented on src/app/core/config/models/config-access-condition-option.model.ts:
```diff
@@ -0,0 +1,8 @@
+export class AccessConditionOption {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/config/models/config-submission-uploads.model.ts:
```diff
@@ -0,0 +1,21 @@
+import {SubmissionFormsModel} from './config-submission-forms.model';
+
+@inheritSerialization(ConfigObject)
+export class SubmissionUploadsModel extends ConfigObject {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/config/submission-uploads-config.service.ts:
```diff
@@ -0,0 +1,18 @@
+import { HALEndpointService } from '../shared/hal-endpoint.service';
+
+@Injectable()
+export class SubmissionUploadsConfigService extends ConfigService {
```
Please add TypeDoc/comments to describe this new service

tdonohue commented on src/app/core/data/request.models.ts:
```diff
@@ -230,7 +231,87 @@ export class IntegrationRequest extends GetRequest {
   }
 }
+
+export class SubmissionFindAllRequest extends GetRequest {
```
Please add TypeDoc/comments to describe *all the new classes in this file* (there are several more in this same file that have no comments to describe them.

tdonohue commented on src/app/core/eperson/eperson-response-parsing.service.ts:
```diff
@@ -0,0 +1,43 @@
+import { ResourceType } from '../shared/resource-type';
+
+@Injectable()
+export class EpersonResponseParsingService extends BaseResponseParsingService implements ResponseParsingService {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/eperson/eperson.service.ts:
```diff
@@ -0,0 +1,20 @@
+import { NormalizedObject } from '../cache/models/normalized-object.model';
+import { DataService } from '../data/data.service';
+
+export abstract class EpersonService<TNormalized extends NormalizedObject, TDomain> extends DataService<TNormalized, TDomain> {
```
Please add TypeDoc/comments to describe this new service

tdonohue commented on src/app/core/eperson/group-eperson.service.ts:
```diff
@@ -0,0 +1,49 @@
+import { PaginatedList } from '../data/paginated-list';
+
+@Injectable()
+export class GroupEpersonService extends EpersonService<NormalizedGroup, Group> {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/integration/models/authority.value.ts:
```diff
@@ -0,0 +1,33 @@
+import { PLACEHOLDER_PARENT_METADATA } from '../../../shared/form/builder/ds-dynamic-form-ui/models/relation-group/dynamic-relation-group.model';
+import { OtherInformation } from '../../../shared/form/builder/models/form-field-metadata-value.model';
+
+export class AuthorityValue extends IntegrationModel {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/json-patch/builder/json-patch-operation-path-combiner.ts:
```diff
@@ -0,0 +1,45 @@
+  path: string;
+}
+
+export class JsonPatchOperationPathCombiner  {
```
In addition to Art's comments, I'll also note that this new class needs TypeDoc/comments to describe it.  If it *is* needed, then we should describe what it is doing that cannot be done by URLCombiner.

atarix83 replied:
I've refactored it using URLCombiner

tdonohue commented on src/app/core/json-patch/builder/json-patch-operations-builder.ts:
```diff
@@ -0,0 +1,111 @@
+import { FormFieldLanguageValueObject } from '../../../shared/form/builder/models/form-field-language-value.model';
+
+@Injectable()
+export class JsonPatchOperationsBuilder {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/json-patch/json-patch-operations.effects.ts:
```diff
@@ -0,0 +1,22 @@
+} from './json-patch-operations.actions';
+
+@Injectable()
+export class JsonPatchOperationsEffects {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/json-patch/json-patch-operations.reducer.ts:
```diff
@@ -0,0 +1,302 @@
+} from './json-patch-operations.actions';
+import { JsonPatchOperationModel, JsonPatchOperationType } from './json-patch.model';
+
+export interface JsonPatchOperationObject {
```
Please add TypeDoc/comments to describe this new interface (and others in this same file).  I see you only added comments for the "JsonPatchOperationsState" interface, but all interfaces should have similar descriptions.

tdonohue commented on src/app/core/json-patch/json-patch-operations.reducer.ts:
```diff
@@ -0,0 +1,302 @@
+
+const initialState: JsonPatchOperationsState = Object.create(null);
+
+export function jsonPatchOperationsReducer(state = initialState, action: PatchOperationsActions): JsonPatchOperationsState {
```
Please add TypeDoc/comments to describe this new function

tdonohue commented on src/app/core/json-patch/json-patch-operations.reducer.ts:
```diff
@@ -0,0 +1,302 @@
+  }
+}
+
+function addOperationToList(body: JsonPatchOperationObject[], actionType, targetPath, value?) {
```
Please add TypeDoc/comments to describe this new function

tdonohue commented on src/app/core/json-patch/json-patch-operations.service.ts:
```diff
@@ -0,0 +1,129 @@
+} from './json-patch-operations.actions';
+import { JsonPatchOperationModel } from './json-patch.model';
+
+export abstract class JsonPatchOperationsService<ResponseDefinitionDomain, PatchRequestDefinition extends PatchRequest> {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/json-patch/json-patch-operations.service.ts:
```diff
@@ -0,0 +1,129 @@
+    return isNotEmpty(resourceID) ? `${endpoint}/${resourceID}` : `${endpoint}`;
+  }
+
+  public jsonPatchByResourceType(linkName: string, scopeId: string, resourceType: string): Observable<ResponseDefinitionDomain> {
```
Please add TypeDoc/comments to all your public methods in this file.  Ideally, all methods should have a brief description, if possible.

tdonohue commented on src/app/core/json-patch/json-patch.model.ts:
```diff
@@ -0,0 +1,14 @@
+export enum JsonPatchOperationType {
```
Please add TypeDoc/comments to describe this new enum (and the class below it)

tdonohue commented on src/app/core/json-patch/selectors.ts:
```diff
@@ -0,0 +1,34 @@
+import { coreSelector, CoreState } from '../core.reducers';
+import { JsonPatchOperationsEntry, JsonPatchOperationsResourceEntry } from './json-patch-operations.reducer';
+
+export function keySelector<T, V>(parentSelector: Selector<any, any>, subState: string, key: string): MemoizedSelector<T, V> {
```
Please add TypeDoc/comments to describe all the new functions in this file

tdonohue commented on src/app/core/shared/file.service.ts:
```diff
@@ -0,0 +1,35 @@
+import { DSpaceRESTV2Response } from '../dspace-rest-v2/dspace-rest-v2-response.model';
+
+@Injectable()
+export class FileService {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/shared/license.model.ts:
```diff
@@ -0,0 +1,14 @@
+import { DSpaceObject } from './dspace-object.model';
+
+export class License extends DSpaceObject {
```
Please add TypeDoc/comments to describe this new class

tdonohue commented on src/app/core/shared/submit-data-response-definition.model.ts:
```diff
@@ -0,0 +1,5 @@
+import { ConfigObject } from '../config/models/config.model';
+import { SubmissionObject } from '../submission/models/submission-object.model';
+
+export type SubmitDataResponseDefinitionObject
```
Please add TypeDoc/comments to describe this new type

atarix83 commented on src/app/core/data/request.service.ts:
```diff
@@ -68,12 +70,28 @@ export class RequestService {
+      });
+  }
+
+  // TODO to review "forceBypassCache" param when https://github.com/DSpace/dspace-angular/issues/217 will be fixed
```
Unfortunately it still doesn't work, request are made but response object is not updated in the store with the new REST response. I suggest to merge as is and open a new task to check if there is a way to not force bypass the cache

atarix83 commented on src/app/submission/edit/submission-edit.component.ts:
```diff
@@ -0,0 +1,72 @@
+      flatMap((params: ParamMap) => this.submissionService.retrieveSubmission(params.get('id')))
+    ).subscribe((submissionObject: SubmissionObject) => {
+      // NOTE new submission is retrieved on the browser side only
+      if (isNotNull(submissionObject)) {
```
Null value is returned on SSR, so it's a control to no perform any operation.
I've changed it to a filter now

atarix83 commented on src/app/submission/sections/container/section-container.component.scss:
```diff
@@ -0,0 +1,22 @@
+@import '../../../../styles/variables';
+
+:host /deep/ .card {
+  margin-bottom: $border-radius * 2;
```
fixed

artlowel replied:
It isn't fixed, you moved it to its own variable and [that is still based on the border radius](https://github.com/DSpace/dspace-angular/pull/279/files#diff-ade74f2136fdbe405029e30cefaec31cR29). Like I said:

>if someone sets the border-radius to 0, because they want hard corners, would you want the margin here to disappear?

Please base it on something related to margins or spacing like [`$spacer`](https://github.com/twbs/bootstrap/blob/v4-dev/scss/_variables.scss#L129), [`$paragraph-margin-bottom`](https://github.com/twbs/bootstrap/blob/v4-dev/scss/_variables.scss#L182), [`$card-spacer-y`](https://github.com/twbs/bootstrap/blob/v4-dev/scss/_variables.scss#L815), etc. any of those would be more suitable than `$border-radius`

atarix83 commented on src/app/submission/edit/submission-edit.component.ts:
```diff
@@ -0,0 +1,72 @@
+
+  ngOnInit() {
+    this.subs.push(this.route.paramMap.pipe(
+      tap((params: ParamMap) => this.submissionId = params.get('id')),
```
fixed

atarix83 commented on src/app/submission/edit/submission-edit.component.ts:
```diff
@@ -0,0 +1,72 @@
+  ngOnInit() {
+    this.subs.push(this.route.paramMap.pipe(
+      tap((params: ParamMap) => this.submissionId = params.get('id')),
+      flatMap((params: ParamMap) => this.submissionService.retrieveSubmission(params.get('id')))
```
fixed

atarix83 commented on src/app/submission/submission.service.ts:
```diff
@@ -0,0 +1,307 @@
+  retrieveSubmission(submissionId): Observable<SubmissionObject> {
+    return this.restService.getDataById(this.getSubmissionObjectLinkName(), submissionId).pipe(
+      filter((submissionObjects: SubmissionObject[]) => isNotUndefined(submissionObjects)),
+      first(),
```
fixed

atarix83 commented on src/app/+login-page/login-page.component.ts:
```diff
@@ -1,20 +1,61 @@
+    const authenticated = this.store.select(isAuthenticated);
+    this.sub = observableCombineLatest(queryParamsObs, authenticated).pipe(
+      filter(([params, auth]) => isNotEmpty(params.token) || isNotEmpty(params.expired)),
+      take(1)
```
fixed

artlowel replied:
No, it's still there in the latest version. But this isn't really a problem, it will work just fine. I just mentioned it for future reference.

atarix83 commented on src/app/core/config/models/config.model.ts:
```diff
@@ -1,17 +1,18 @@
 
   @autoserialize
-  public type: string;
+  public type: any;
```
fixed

atarix83 commented on src/app/core/data/data.service.ts:
```diff
@@ -53,6 +57,42 @@ export abstract class DataService<TNormalized extends NormalizedObject, TDomain>
     }
   }
 
+  protected getSearchByHref(endpoint, searchByLink, options: FindAllOptions = {}): Observable<string> {
```
refactored

atarix83 commented on src/app/core/data/data.service.ts:
```diff
@@ -53,6 +57,42 @@ export abstract class DataService<TNormalized extends NormalizedObject, TDomain>
+    let result: Observable<string>;
+    const args = [];
+
+    result = observableOf(`${endpoint}/search/${searchByLink}`);
```
fixed

atarix83 commented:
@artlowel I've made changes based on your review, you want you can proceed with the second round.

@tdonohue I've checked that deposit doesn't work because workflowitem REST endpoint is still missing to the current version of REST server.
I will add TypeDocs asap

artlowel commented on src/styles/_custom_variables.scss:
```diff
@@ -22,3 +25,5 @@ $admin-sidebar-header-bg: darken($dark, 7%);
 $dark-scrollbar-background: $admin-sidebar-active-bg;
 $dark-scrollbar-foreground: #47495d;
+
+$submission-sections-margin-bottom: $border-radius * 2;
```
See my [comment above](https://github.com/DSpace/dspace-angular/pull/279#discussion_r252595273)

artlowel commented on src/app/shared/chips/chips.component.ts:
```diff
@@ -79,16 +85,34 @@ export class ChipsComponent implements OnChanges {
 
       this.cdr.detectChanges();
-      tooltip.open();
+      if (!chipsItem.hasIcons() || (chipsItem.hasIcons() && !chipsItem.hasVisibleIcons()) || field) {
```
I don't think the second `chipsItem.hasIcons()` is necessary. If `!chipsItem.hasIcons()` is false (and therefore we move to the second part of the OR), `chipsItem.hasIcons()` must be true

atarix83 commented:
@tdonohue @artlowel typedoc added and made changes based on feedback

artlowel commented on src/app/submission/sections/license/section-license.component.ts:
```diff
@@ -0,0 +1,156 @@
+          ).pipe(
+            take(1),
+            filter((isReadOnly) => isReadOnly))
+            .subscribe(() => {
```
This subscribe is never unsubscribed. It is also nested inside another subscribe, so it can probably be rewritten using switchMap

atarix83 replied:
fixed

artlowel commented on src/app/submission/sections/license/section-license.component.ts:
```diff
@@ -0,0 +1,156 @@
+      this.sectionService.getSectionErrors(this.submissionId, this.sectionData.id).pipe(
+        filter((errors) => isNotEmpty(errors)),
+        distinctUntilChanged())
+        .subscribe((errors) => {
```
This subscribe is never unsubscribed

atarix83 replied:
fixed

artlowel commented on src/app/submission/sections/sections.service.ts:
```diff
@@ -0,0 +1,195 @@
+          if (!enabled) {
+            this.translate.get('submission.sections.general.metadata-extracted-new-section', {sectionId})
+              .pipe(take(1))
+              .subscribe((m) => {
```
Nested subscribes can likely be rewritten using switchMap.

atarix83 replied:
please take a look to this stackoverflow answer https://stackoverflow.com/questions/40861494/angular2-unsubscribe-from-http-observable-in-service
there is no need to unsubscribe in a service

artlowel replied:
You're right about unsubscribing in services. However, I commented on the fact that there was a subscribe inside another subscribe. This is an anti-pattern, and it can be rewritten using switchMap, mergeMap or flatMap. More info here:
- https://blog.strongbrew.io/rxjs-best-practices-in-angular/#avoiding-nested-subscribes
- https://medium.com/@paynoattn/3-common-mistakes-i-see-people-use-in-rx-and-the-observable-pattern-ba55fee3d031#ad2b
- http://www.syntaxsuccess.com/viewarticle/fixing-rxjs-anti-patterns

atarix83 replied:
removed nested subscription

artlowel commented on src/app/submission/sections/upload/file/file.component.ts:
```diff
@@ -0,0 +1,194 @@
+      this.submissionId,
+      this.pathCombiner.rootElement,
+      this.pathCombiner.subRootElement)
+      .subscribe();
```
This subscribe is never unsubscribed. It also doesn't have any contents, does it need to be there?

atarix83 replied:
fixed

atarix83 commented on src/app/shared/date.util.ts:
```diff
@@ -1,5 +1,12 @@
 
+import { isObject } from 'lodash';
+
+export function isDateObject(value) {
```
Renamed method

atarix83 commented on src/app/shared/date.util.ts:
```diff
@@ -1,5 +1,12 @@
+    && value.hasOwnProperty('month')  && value.hasOwnProperty('year');
+}
+
 export function dateToGMTString(date: Date | NgbDateStruct) {
```
refactored method using moment.js

atarix83 commented on src/app/shared/form/builder/ds-dynamic-form-ui/models/lookup/dynamic-lookup.component.ts:
```diff
@@ -180,43 +199,67 @@ export class DsDynamicLookupComponent extends DynamicFormControlComponent implem
+        return observableOf(emptyResult);
+      }),
+      distinctUntilChanged())
+      .subscribe((object: IntegrationData) => {
```
fixed

atarix83 commented on src/app/submission/form/collection/submission-form-collection.component.ts:
```diff
@@ -0,0 +1,187 @@
+            find((collections: RemoteData<PaginatedList<Collection>>) => !collections.isResponsePending && collections.hasSucceeded),
+            mergeMap((collections: RemoteData<PaginatedList<Collection>>) => collections.payload.page),
+            filter((collectionData: Collection) => isNotEmpty(collectionData)),
+            tap((collectionData: Collection) => {
```
fixed

atarix83 commented on src/app/submission/form/submission-upload-files/submission-upload-files.component.ts:
```diff
@@ -0,0 +1,100 @@
+      this.submissionService.getSubmissionObjectLinkName(),
+      this.submissionId,
+      'sections')
+      .subscribe();
```
fixed

atarix83 commented:
@artlowel 
I've made changes based on your feedback.

About issues you reported : 

> Upload
This is due to a limit of uploaded files size on REST side. We will increase limit

> 2nd describe panel issue
due to a wrong submission-form configuration. Fixed

> Deposit
Deposit doesn't work because workflow endpoint is missing on REST demo server at the moment

> Upload files
The problem was with `Content-Disposition` header, where the file name is retrieved. to be able to retrieve it from angular side, we need to set 'Access-Control-Expose-Headers': 'Content-Disposition' ON SERVER SIDE. Fixed

> Autocompleting dropdown
This is due to an issue with pagination of authority endpoint. Fix will be include in workflow REST PR

> Add More button
disabled when no options are available

atarix83 commented on resources/i18n/en.json:
```diff
@@ -485,5 +499,84 @@
+      "upload": {
+        "no-entry": "No",
+        "no-file-uploaded": "No file uploaded yet.",
+        "info": "Here you will find all the files currently in the item. You can update the fle metadata and access conditions or <strong>upload additional files just dragging&dropping them everywhere in the page</strong>",
```
fixed

atarix83 commented:
@artlowel 
> I can't figure out what the edit icon next to a file in the "Uploaded files" panel does. If I click it, the file name and description disappear, but I can't type in anything to replace them. They reappear and look unchanged if I cancel editing.

this is how it should work

![chrome-capture 2](https://user-images.githubusercontent.com/2486489/53191519-50865880-360c-11e9-95f0-b1e814d621fc.gif)

isn't it for you?

artlowel commented:
Thanks @atarix83!

Most of the UI issues I reported have been fixed, either through code changes or rest config. The deposit issue should be fixed once https://github.com/DSpace/DSpace/pull/2312 is merged. I still think it's strange that you can't type to search in those autocompleting dropdowns for e.g. type and language, but that is a minor usability gripe.

Most of the code issues have been fixed as well, the [ones](https://github.com/DSpace/dspace-angular/pull/279#discussion_r211564360) that [remain](https://github.com/DSpace/dspace-angular/pull/279#discussion_r211581279) are quite minor, and the PR can be merged without them being addressed. 

The main thing still lacking is typedocs. There are quite big parts of this PR with missing comments. Please don't make @tdonohue list them all 😉 

Automatic metadata extraction is disabled on the default demo endpoint through config, but when testing against the [or2018 endpoint](https://dspace7.4science.it/or2018/) I noticed the automatic metadata extraction still happens without user interaction. @tdonohue made a [jira ticket](https://jira.duraspace.org/browse/DS-4002) about this issue, and if I recall correctly it will require a change on the rest side before it can be addressed in angular, but we can't forget about it.

paulo-graca commented:
I'm still reviewing this PR, but I just want to let you know 2 things:
- As occurred with @artlowel I'm getting this error too with the second section: 
![2019-02-28 14_43_44-edit submission](https://user-images.githubusercontent.com/877488/53574386-68f3f700-3b67-11e9-92ab-3efd9eb7dcb7.png)
- and the other thing, If you play around with `/workspaceitems/?/edit` address, replacing "?" with any number, even with unauthorized users, you will be able to access that content. This is a critical security issue. You shouldn't be able to see others submissions.

atarix83 commented:
@paulo-graca 
>  As occurred with @artlowel I'm getting this error too with the second section

this problem is due to a wrong submission-form configuration that I have to manually fix every time a  Demo REST server is deployed. It will be finally fixed with https://github.com/DSpace/DSpace/pull/2312

> and the other thing, If you play around with `/workspaceitems/?/edit` address, replacing "?" with any number, even with unauthorized users, you will be able to access that content. This is a critical security issue. You shouldn't be able to see others submissions.

You are right, this should be fixed on REST side. however you should not be able to save a submission with unauthorized users. with last commit you should display an error notification when error occurred on save

paulo-graca commented:
My problem with the submission: I'm getting a 404 error on the POST Request URL: http://dev5.rcaap.pt/spring-rest/api/workflowitems.

atarix83 commented:
@paulo-graca 
thanks for your review.

> My problem with the submission: I'm getting a 404 error on the POST Request URL: http://localhost/spring-rest/api/workflowitems.

The deposit of the submission will work once [DSpace/DSpace#2312](https://github.com/DSpace/DSpace/pull/2312) is merged

artlowel commented on src/app/submission/form/collection/submission-form-collection.component.ts:
```diff
@@ -0,0 +1,193 @@
+
+      // @TODO replace with search/top browse endpoint
+      // @TODO implement community/subcommunity hierarchy
+      const communities$ = this.communityDataService.findAll().pipe(
```
Am I right in the assumption that this would only show the 20 (the first page) of communities, and for each community the first 20 collections. So, that as it is now, you can't submit to a collection that isn't on that first page?

I ask because @LotteHofstede  faced the same problem when designing the context sensitive admin menu popups, and as far as I can tell, it can't be done yet with the rest api as it is now (short of fetching _all_ communities and collections and searching the tree on the client)

atarix83 replied:
you are right, I was deceived by FindAll method. If you are agree we could resolve this with a separate issue, maybe when we will have a rest endpoint for this feature

artlowel replied:
I agree to solve this separately, that's why I suggested that search-as-you-type rest endpoint in the last meeting

For now, that collection search can still work if you simply query the discovery endpoint directly and filter by dso type. [Here](https://github.com/DSpace/dspace-angular/pull/367/files#diff-eacbb9b4341c4775a53c58ad82a90870R84) is an example from @LotteHofstede's PR. This has two downsides: 
- You need to type whole words before the discovery endpoint returns results. 
- You also can't show the parent communities, because those come with their child collections embedded, so we end up downloading way too much to simply show those parents. I'd reconsider that, if we have support for partial objects.

So the usability leaves some things to be desired, but it will at least work with all collections.

atarix83 replied:
created an [issue](https://github.com/DSpace/dspace-angular/issues/373) for this

atarix83 commented on src/app/submission/sections/upload/file/file.component.html:
```diff
@@ -0,0 +1,68 @@
+      <div class="float-left w-75">
+        <h3>{{fileName}} <span class="text-muted">({{fileData?.sizeBytes | dsFileSize}})</span></h3>
+      </div>
+      <div class="float-right w-15" [class.sticky-buttons]="!readMode">
```
this is a design choice, with default size of icon they are too small

atarix83 commented on src/app/+login-page/login-page.component.ts:
```diff
@@ -1,20 +1,61 @@
   templateUrl: './login-page.component.html'
 })
-export class LoginPageComponent implements OnDestroy {
+export class LoginPageComponent implements OnDestroy, OnInit {
```
done

atarix83 commented on src/app/shared/alerts/alerts.component.ts:
```diff
@@ -0,0 +1,44 @@
+import { fadeOutLeave, fadeOutState } from '../animations/fade';
+
+@Component({
+  selector: 'ds-alert',
```
Renamed

atarix83 commented on src/app/shared/alerts/alerts.component.ts:
```diff
@@ -0,0 +1,44 @@
+import { ChangeDetectorRef, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
```
Added

atarix83 commented on src/app/submission/form/submission-form.component.ts:
```diff
@@ -0,0 +1,131 @@
+import { SubmissionObject } from '../../core/submission/models/submission-object.model';
+
+@Component({
+  selector: 'ds-submission-submit-form',
```
renamed component properly

paulo-graca replied:
for a coherence matter, could the selector be renamed to: ds-submission-form ?

atarix83 commented on src/app/submission/sections/container/section-container.component.ts:
```diff
@@ -0,0 +1,49 @@
+import { AlertType } from '../../../shared/alerts/aletrs-type';
+
+@Component({
+  selector: 'ds-submission-form-section-container',
```
renamed component properly

atarix83 commented on src/app/submission/sections/form/section-form.component.ts:
```diff
@@ -0,0 +1,237 @@
+import { difference } from '../../../shared/object.util';
+
+@Component({
+  selector: 'ds-submission-section-form',
```
renamed component properly

atarix83 commented on src/app/submission/sections/license/section-license.component.ts:
```diff
@@ -0,0 +1,154 @@
+import { FormComponent } from '../../../shared/form/form.component';
+
+@Component({
+  selector: 'ds-submission-section-license',
```
renamed component properly

atarix83 commented on src/app/submission/sections/upload/section-upload.component.ts:
```diff
@@ -0,0 +1,206 @@
+export const POLICY_DEFAULT_WITH_LIST = 2; // Banner2
+
+@Component({
+  selector: 'ds-submission-section-upload',
```
renamed component properly

atarix83 commented on src/app/submission/submit/submission-submit.component.ts:
```diff
@@ -0,0 +1,68 @@
+import { Collection } from '../../core/shared/collection.model';
+
+@Component({
+  selector: 'ds-submit-page',
```
renamed properly

atarix83 commented:
@paulo-graca @tdonohue @artlowel 

I've made changes requested by @paulo-graca 

I've added more comments on code, I think that now the code is quite commented.

I look forward to your final approval. thanks

tdonohue commented:
@atarix83 : Just a quick note. It looks like some new conflicts were just created in this PR (as I just merged #367, which was approved).  Hopefully these will be minor to resolve.

atarix83 commented:
@tdonohue conflicts resolved

paulo-graca commented on src/app/shared/form/builder/ds-dynamic-form-ui/ds-dynamic-form-control-container.component.ts:
```diff
@@ -0,0 +1,215 @@
+}
+
+@Component({
+  selector: 'ds-dynamic-form-control',
```
could you change this selector to: ds-dynamic-form-control-container

paulo-graca commented on src/app/shared/form/builder/ds-dynamic-form-ui/models/date-picker-inline/dynamic-date-picker-inline.component.ts:
```diff
@@ -0,0 +1,35 @@
+} from '@ng-dynamic-forms/core';
+
+@Component({
+  selector: 'ds-date-picker-inline',
```
could you change this selector to: ds-dynamic-date-picker-inline ?

paulo-graca commented on src/app/submission/sections/container/section-container.component.ts:
```diff
@@ -0,0 +1,93 @@
+ * This component represents a section that contains the submission license form.
+ */
+@Component({
+  selector: 'ds-submission-section-container',
```
This selector name isn't coherent with others.

atarix83 replied:
I think that this follows the same rule used for other sections components

atarix83 commented:
Thank you @paulo-graca for reviewing again
I've added missing comments, except for all DynamicForms components. I think that also other DynamicForms components that you have not reported needs more comments, so if @tdonohue is agree I'll open a separate issue for this.

I've also renamed component selectors that you reported except for the one I've replied to

artlowel commented:
I tested it again, and quickly went over the changed code since my last review. Everything still works, and most comments are present as well. Thanks for the hard work @atarix83!

Let's discuss what to do with the DynamicForms comments in the meeting today. If we decide to merge it as is, I'll approve the PR.

tdonohue commented on src/app/pagenotfound/pagenotfound.component.ts:
```diff
@@ -1,14 +1,20 @@
-export class PageNotFoundComponent {
-  constructor(responseService: ServerResponseService) {
-    responseService.setNotFound();
+export class PageNotFoundComponent implements OnInit {
```
We can let this one slide. It's a component that already existed.  It does need (eventual) comments, but is not required of this specific PR.

tdonohue commented:
@atarix83 : Sorry, I just created a minor merge conflict with yarn.lock after merging #375 (Bootstrap upgrade).  Could you do a quick update to this PR?

atarix83 commented:
@tdonohue 

I've resolved conflicts and fixes some little issue due to bootstrap update

atarix83 commented:
@tdonohue @artlowel @paulo-graca 
I've created issue for DynamicForms comments [here](https://github.com/DSpace/dspace-angular/issues/378)

tdonohue commented:
@atarix83 : Apologies, but I again seem to have created some minor merge conflicts here (likely from merging #374).  This PR is the *next one* to merge though.  So, if you could clean it up one last time, we can get it merged.

atarix83 commented:
@tdonohue  conflicts are resolved now, you can merge it. thanks

tdonohue commented:
Also thanks for the code reviews from @artlowel and @paulo-graca !  I know this was a massive effort all around :)
